MREUnityRuntime/MREUnityRuntimeLib/Core/Physics/JitterBuffer.cs

```diff
@@ -77,390 +102,815 @@ public Snapshot(float time, List<TransformInfo> transforms)
+			float nextTimeStamp = caclNextTimeStamp(timestep);
+
+			if (!_wasNetworkHealthy && _isNetworkHealthy)
+			{
+				nextTimeStamp = Math.Min(nextTimeStamp, LastSnapshotLocalTime);
+			}
+
+			// if network is healthy, we don't want large buffer time
+			if (_isNetworkHealthy)
+			{
+				const float maxBufferTime = 0.1f;
+				nextTimeStamp = Math.Max(nextTimeStamp, LastSnapshotLocalTime - maxBufferTime);
+			}
+
+			// Update state from available snapshots
+			stepBufferAndUpdateRigidBodies(nextTimeStamp);
+
+			_wasNetworkHealthy = _isNetworkHealthy;
 		}
 
-		public SortedList<Guid, RigidBodyState> RigidBodies = new SortedList<Guid, RigidBodyState>();
-	}
+		private enum Action
+		{
+			Add,
+			Remove
+		}
 
-	/// <summary>
-	/// Multi-source time and snapshot manager.
-	/// </summary>
-	public class TimeSnapshotManager
-	{
-		public class SourceInfo
+		public void RegisterRigidBody(Guid id)
 		{
-			/// <todo>
-			/// is this required?
-			/// </todo>
-			private enum Mode
+			_pendingRigidBodyManagementActions[id] = Action.Add;
+		}
+
+		public void UnregisterRigidBody(Guid id)
+		{
+			_pendingRigidBodyManagementActions[id] = Action.Remove;
+		}
+
+		#endregion
+
+		#region Stats
+
+		bool _isNetworkHealthy = true;
+		bool _wasNetworkHealthy = true;
+
+		/// <summary>
+		/// Based on time covered by updates received with last second.
+		/// </summary>
+		class UpdateTimeHealth
+		{
+			/// <summary>
+			/// Running average time covered with updates within past second.
+			/// </summary>
+			public float UpdatedTime = 1.0f;
+
+			private int _runningWindowsSize = 60;
+
+			public void addSample(float updateTime)
 			{
-				Init,
-				Play,
+				UpdatedTime -= UpdatedTime / _runningWindowsSize;
+				UpdatedTime += updateTime;
 			}
+		}
 
-			/// <todo>
-			/// add comments
-			/// </todo>
-			class RunningStats
-			{
-				int _count = 0;
+		/// <summary>
+		/// Running average for update availability at different time checkpoints relative to current output.
+		/// </summary>
+		class RunningStats
+		{
+			const int _numRunningWindowSamples = 100;
+
+			public const float TimeStep = 1.0f / 120;
 
-				float _mean = 0.0f;
-				float _variance = 0.0f;
+			const float _timespan = 0.1f;
 
-				// todo: expose as a parameter
-				int _windowSize = 120;
+			const int _count = (int)(2 * _timespan / TimeStep) + 1;
 
-				public void shiftTime(float time)
+			const int _current = _count / 2;
+
+			private float[] _bufferTimeHeuristics = new float[_count];
+
+			/// <summary>
+			/// Estimated update availability if output is slown down.
+			/// </summary>
+			public float SlowDownIndicator { get { return _bufferTimeHeuristics[_current + 1]; } }
+
+			/// <summary>
+			/// Estimated update availability with current output rate.
+			/// </summary>
+			public float CurrentRateQuality { get { return _bufferTimeHeuristics[_current]; } }
+
+			/// <summary>
+			/// Estimated update availability if output is sped down.
```

Comment:
slowed

Reply:
sped up

---
In reply to: [450948418](https://github.com/microsoft/mixed-reality-extension-unity/pull/210#discussion_r450948418) [](ancestors = 450948418)